### PR TITLE
Authenticate using manually obtained refresh token

### PR DESCRIPTION
**Pull request by eduardog3000**
Confirmed working with Apple sign-in. I can't test the others but this should be provider agnostic.

Manually intercept the refresh token using something like [HTTP Toolkit](https://httptoolkit.com/) then use it as the sole argument to the constructor.

### CODE REVIEW

**Comment by jhansche** [comment]
I'm not against allowing the access or refresh token to be specified manually. I'm just not seeing the usefulness either. But here are a few issues that come to mind:

* This isn't really the same as "allowing SSO" - which would involve making the correct GraphQL call with a provider token, in exchange for the access and refresh tokens. Again, not against allowing this, but we should call it what it is: allow setting the refresh token manually.
* This solution doesn't work on Android unfortunately, due to certificate pinning. The release version of the app does not allow intercepting the SSL traffic, even with user-installed certificates. Without that, you can't obtain this token using the tool you suggest. I'm surprised it worked for you on iOS, because on Android the certificate pinning is done in the underlying React Native code, which is commonly shared between platforms.
* When used with the HA integration, this email field won't be redacted.

I think rather than taking the token in the email argument, I'd prefer we switch that to a keyword argument, allowing email and password to default to None, and specify the token(s) as keyword args.
```
def __init__(self, email=None, password=None, /, access_token=None, refresh_token=None)
```

**Comment by eduardog3000** [comment]
> I'm just not seeing the usefulness either.

Enabling people with SSO accounts to use this at all, even if it takes some manual work and isn't perfect.

* I guess "bypass SSO" would be more accurate
* I hadn't heard of that. I will note I wasn't able to intercept the actual sign in process, but I chalked that up to Apple making their process more secure. I just intercepted requests after signing in until it had to refresh its access token. I'd test on a Pixel I have, but it looks like the app doesn't implement Apple SSO on Android ([even though it's entirely possible](https://cloud.google.com/identity-platform/docs/android/apple)).

> In this case, if your refresh token becomes invalid, will it fail in a meaningful way?

I guess not. But I don't particularly see how that's any different from [what it already does when the refresh fails](https://github.com/jhansche/pybirdbuddy/blob/4706d0b336e4d25c68f8157bd0fbff5508d6381e/birdbuddy/client.py#L146) or where that falls back to email and password afterward.

> When used with the HA integration, this email field won't be redacted.
> I'd prefer we switch that to a keyword argument, allowing email and password to default to None, and specify the token(s) as keyword args.

Ahh yeah, my Python is a little rusty. That's a much better idea.

**Comment by jhansche** [comment]
> Enabling people with SSO accounts to use this at all, even if it takes some manual work and isn't perfect.

Yeah, I agree that being able to use it at all is more helpful than not being able to :)

My hesitation is that my idea of properly supporting SSO involves using the HA OAuth application credentials flow, in order to show a web view that prompts to authorize the access, returning the (Google | Apple | Facebook) oauth access token, which we then use in client.BirdBuddy to call the `authSocialSignIn` graphql endpoint, which is going to return our BirdBuddy access_token and refresh_token.

So while this may enable partial access, it is not actually related to "SSO" in any way.

> I will note I wasn't able to intercept the actual sign in process.

Correct, I would not expect to be able to intercept the process between Bird Buddy App <-> Apple Sign-in.

What happens is after that process between app and Apple, Apple returns an oauth token of some sort to the app, and then the app sends that token to the `authSocialSignIn` endpoint. The graphql server processes that token, validates the token with the provider (Apple/etc), and uses that to authenticate the user. Once authenticated, it returns the exact same access_token and refresh_token that we get from the `emailSignIn` endpoint with email+password.

> I guess not. But I don't particularly see how that's any different from what it already does

You're right - what I'm talking about is that it would "fall back to email" afterward, make [this request](https://github.com/jhansche/pybirdbuddy/blob/4706d0b336e4d25c68f8157bd0fbff5508d6381e/birdbuddy/client.py#L115-L123) with email=null, password=null, and then fail.

> I just intercepted requests after signing in until it had to refresh its access token.

Right, this is the part that surprises me. On Android, it's not possible to intercept _any_ of the HTTP calls between the Bird Buddy app and the graphql endpoint. That's because the React Native code that wraps the GraphQL communication also enforces the certificate pinning of that endpoint. If we try to intercept that HTTP traffic, then the SSL certificate of our mitm proxy will not match the certificate they expect, and it will fail the handshake.

The only way I was able to monitor this HTTP traffic was by making the Android app debuggable, and then attaching the Android Studio profiler, where I can monitor HTTP requests. But in doing so, I lose the ability for the app to make the Sign-In authorization request with Google.

File: birdbuddy/client.py [hunk 1]
```diff
@@ -40,21 +40,27 @@ class BirdBuddy:
...
+    _password: str | None
     _access_token: str | None
     _refresh_token: str | None
     _me: BirdBuddyUser | None
     _feeders: dict[str, Feeder]
     _collections: dict[str, Collection]
     _last_feed_date: datetime
 
-    def __init__(self, email: str, password: str) -> None:
+    def __init__(self, emailOrToken: str, password: str | None = None) -> None:
```
Comment (jhansche) [inline birdbuddy/client.py:52]:
```
def __init__(
    self,
    email: str | None = None,
    password: str | None = None,
    /,
    access_token: str | None = None,
    refresh_token: str | None = None,
) -> None:
    # then replace lines 56-57 to assign their values from the kwargs
```

Then we can use it like this:
```
client = BirdBuddy(refresh_token="your manual token")
await client.refresh()
```

**Review by jhansche — CHANGES_REQUESTED** [review]

File: birdbuddy/client.py (hunk 1 above)
Comment (jhansche) [reply birdbuddy/client.py]: I made these changes and also merged in master.

Mind testing this change using your manual refresh token method, like shown above?

Comment (eduardog3000) [reply birdbuddy/client.py]: Yep, it works.

Comment (jhansche) [reply birdbuddy/client.py]: Great! I'll get this merged, thanks!

**Pull request merged**